### PR TITLE
[SPARK-32002][SQL]Support ExtractValue from nested ArrayStruct

### PR DESCRIPTION
**Pull request by guiyanakuang**
### What changes were proposed in this pull request?

For data nest.json

```json
{"a": [{"b": [{"c": [1,2]}]}]}
{"a": [{"b": [{"c": [1]}, {"c": [2]}]}]}
```

run with

```scala
val df: DataFrame = spark.read.json(testFile("nest-data.json"))
df.createTempView("nest_table")
sql("select a.b.c from nest_table").show()
```

will got error

```log
org.apache.spark.sql.AnalysisException: cannot resolve 'nest_table.`a`.`b`['c']' due to data type mismatch: argument 2 requires integral type, however, ''c'' is of string type.; line 1 pos 7;
'Project [a#6.b[c] AS c#8|#6.b[c] AS c#8]
+- SubqueryAlias `nest_table`
+- Relationa#6 json
```

Analyse the causes, a.b Expression dataType match extractor for c, but a.b extractor is GetArrayStructFields, ArrayType(ArrayType()) match GetArrayItem, extraction ("c") treat as an ordinal.

### Why are the changes needed?

Spark sql cannot analyse nested arrays, it is very common to analyse this type of data, especially in the field of advertising!

### Does this PR introduce _any_ user-facing change?

Users can query nested arrays directly using sql and pruning is supported.

### How was this patch tested?

Added UT

ComplexTypesSuite 
Added tests show that querying nested arrays is possible.

NestArraySchemaPruningSuite
Test extraction of nested arrays while supporting schema pruning.

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/ComplexTypesSuite.scala [hunk 1]
```diff
@@ -117,4 +116,13 @@ class ComplexTypesSuite extends QueryTest with SharedSparkSession {
...
   }
+
+  test("SPARK-32002: Support ExtractValue from nested ArrayStruct") {
+    val jsonStr1 = """{"a": [{"b": [{"c": [1,2]}]}]}"""
+    val jsonStr2 = """{"a": [{"b": [{"c": [1]}, {"c": [2]}]}]}"""
+    val df = spark.read.json(Seq(jsonStr1, jsonStr2).toDS())
+    checkAnswer(df.select($"a.b.c"), Row(Seq(Seq(Seq(1, 2))))
+      :: Row(Seq(Seq(Seq(1), Seq(2)))) :: Nil)
+  }
+
```
Comment (AngersZhuuuu) [inline sql/core/src/test/scala/org/apache/spark/sql/ComplexTypesSuite.scala:127]: Can you add more test case show more depth such as `a.b.c.d` etc..

File: sql/core/src/test/scala/org/apache/spark/sql/ComplexTypesSuite.scala [hunk 2]
```diff
@@ -18,11 +18,10 @@
 package org.apache.spark.sql
 
 import scala.collection.JavaConverters._
-
 import org.apache.spark.sql.catalyst.expressions.CreateNamedStruct
```
Comment (AngersZhuuuu) [inline sql/core/src/test/scala/org/apache/spark/sql/ComplexTypesSuite.scala:21]: nit:unnecessary change

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeExtractors.scala [hunk 3]
```diff
@@ -218,6 +226,77 @@ case class GetArrayStructFields(
...
+  }
+}
+
+case class ExtractNestedArrayField(
+  child: Expression,
+  ordinal: Int,
+  numFields: Int,
+  field: StructField,
+  containsNullSeq: Seq[Boolean]) extends UnaryExpression
+  with ExtractValue with NullIntolerant with CodegenFallback {
```
Comment (AngersZhuuuu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeExtractors.scala:254]: Nit:  Indentation is 4

File: sql/core/src/test/scala/org/apache/spark/sql/ComplexTypesSuite.scala (hunk 1 above)
Comment (guiyanakuang) [reply sql/core/src/test/scala/org/apache/spark/sql/ComplexTypesSuite.scala:127]: OK, I add a test for which the depth can be adjusted.

**Comment by HyukjinKwon** [comment]
ok to test

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeExtractors.scala [hunk 4]
```diff
@@ -218,6 +226,77 @@ case class GetArrayStructFields(
...
 }
 
+object ExtractNestedArray {
+  type ReturnType = Option[(DataType, Boolean, Seq[Boolean])]
+
+  def unapply(dataType: DataType): ReturnType = {
+    extractArrayType(dataType)
+  }
+
+  def extractArrayType(dataType: DataType): ReturnType = {
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeExtractors.scala:236]: @guiyanakuang can you address the comments in https://github.com/apache/spark/pull/28860?

**Comment by HyukjinKwon** [comment]
cc @viirya and @gatorsmile

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeExtractors.scala (hunk 4 above)
Comment (guiyanakuang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeExtractors.scala:236]: Yes, I will modify some of the code and supplementary documentation later.

**Comment by maropu** [comment]
> Spark sql cannot analyse nested arrays, it is very common to analyse this type of data, especially in the field of advertising!

I remember the related discussion in https://github.com/apache/spark/pull/29645 Btw, other systems support this resolution?

**Comment by viirya** [comment]
Hmm, yeah, I remember the discussion with @cloud-fan. IIUC, it is not supported explicitly in Spark.

**Comment by guiyanakuang** [comment]
> I remember the related discussion in #29645 Btw, other systems support this resolution?

Google BigQuery and Amazon Athena are supposed to support nested arrays, but instead of analysing the nested data directly, they use utdf to unfold the data for analysis. Like this

```sql
select  c_column, d_column from table lateral view unnest(a.b.c, a.b.d) as c_column, d_column
```

However, the spark sql syntax analysis phase is exceptional because no extractor for nested arrays could be found.

**Comment by guiyanakuang** [comment]
> Hmm, yeah, I remember the discussion with @cloud-fan. IIUC, it is not supported explicitly in Spark.

Does this mean I need to close the pr?

**Comment by viirya** [comment]
cc @cloud-fan

**Comment by HyukjinKwon** [comment]
@guiyanakuang, can we make it matched the examples you mentioned? Also, just to double check, we don't have any reference for the support nested array support, right?

**Comment by guiyanakuang** [comment]
> @guiyanakuang, can we make it matched the examples you mentioned? Also, just to double check, we don't have any reference for the support nested array support, right?

Are you referring to this example ?
```sql
select  c_column, d_column from table lateral view unnest(a.b.c, a.b.d) as c_column, d_column
```
Yes, we can implement this example, it's just that Unnest utdf does not yet exist for spark, and we can use explore instead.

```scala
val jsonStr1 = """{"a": [{"b": [{"c": [1,2]}]}]}"""
val jsonStr2 = """{"a": [{"b": [{"c": [1]}, {"c": [2]}]}]}"""
val df = spark.read.json(Seq(jsonStr1, jsonStr2).toDS())
df.createTempView("nested_arrays")
sql(
  """
    |select
    |    c_u_u_u
    |from
    |    nested_arrays
    |    lateral view explode(a.b.c) as c_u
    |    lateral view explode(c_u)   as c_u_u
    |    lateral view explode(c_u_u) as c_u_u_u
    |""".stripMargin).show()

+-------+
|c_u_u_u|
+-------+
|      1|
|      2|
|      1|
|      2|
+-------+
```
In fact, my company, Tencent, has implemented unnest in Spark, which is also referenced in Google's BigQuery.
Our advertising analysts rely heavily on it.

**Comment by cloud-fan** [comment]
I don't think this is the right direction, as people will ask for more in the future, like supporting 3-level nested array. At the end, you can't even tell what `a.b` means, as `a` can be an arbitrarily nested array. I like the `unnest` from BigQuery as it has a clear semantic. The `explode` approach also works so I don't see why we need this feature. I'm fine to add `unnest` to Spark as an alternative to solve this use case.

**Comment by guiyanakuang** [comment]
I would like to explain a little more
I also like the clearly defined `unnest`, which addresses this use case in my company
```sql
select c_u_u_u from nested_arrays lateral view unnest(a.b.c) as c_u_u_u
```

This use case is currently solved by spark in this way

```sql
select
    c_u_u_u
from
    nested_arrays 
    lateral view explode(a.b) as c_u 
    lateral view explode(c_u) as c_u_u
    lateral view explode(c_u_u.c) as c_u_u_u
```

The explode approach does also work, but when it comes to solving multiple nested array fields, the writing style is a headache. I know, it is difficult to understand what `a.b/a.b.c` means when `a` is a nested array. The good news is that analysts just need the values of all the leaves and `unnest(a.b.c)` makes it easy for us to solve the problem without having to understand the middle structure. 

**This pr is a prerequisite for my solution**

1. This pr implements the corresponding extractor, which supports nesting at any level, not just 3. I understand that this flexibility can lead to ambiguity and I would not recommend using `a.b.c` directly, but in combination with unnest.

2. This pr can be combined very well with pushdown. In the current way only `a.b` can be extracted using `GetArrayStructFields`, which filters out the `c` brothers in memory. Leads to amplification of read data io. This pr guarantees that only the corresponding data will be read, based on Dremel storage, similar to what can be done with parquet.

**Comment by cloud-fan** [comment]
I'm not sure `a.b` can be pushed down if `a` is a deeply nested array. I'm not against adding a new internal expression to extract field from nested array, if it's needed by `unnest`. I'm -1 to overload `a.b` to support arbitrary levels of nested arrays.

**Comment by guiyanakuang** [comment]
I think it's a good idea to add an internal expression to extract nested arrays, but it's only available to internal functions and not directly exposed. I need some time to adapt the code. Thanks for the advice.  I will close this now.
